src/target/target_flash.c

```diff
@@ -0,0 +1,271 @@
+/*
+ * This file is part of the Black Magic Debug project.
+ *
+ * Copyright (C) 2022  Black Sphere Technologies Ltd.
+ * Copyright (C) 2022  Rafael Silva <perigoso@riseup.net>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This file implements target flash interaction routines.
+ * Provides functionality for buffered flash operations
+ * It depends on target flash implementations
+ */
+
+#include "general.h"
+#include "target_internal.h"
+
+static int flash_buffered_write(target_flash_s *f, target_addr_t dest, const void *src, size_t len);
+static int flash_buffered_flush(target_flash_s *f);
+
+target_flash_s *target_flash_for_addr(target *t, uint32_t addr)
+{
+	for (target_flash_s *f = t->flash; f; f = f->next)
+		if ((f->start <= addr) && (addr < (f->start + f->length)))
+			return f;
+	return NULL;
+}
+
+static int target_enter_flash_mode(target *t)
+{
+	if (t->flash_mode)
+		return 0;
+
+	int ret = 0;
+	if (t->enter_flash_mode)
+		ret = t->enter_flash_mode(t);
```

Comment:
This looks to be boolean ("worked"/"failed") - why not use bool here and have this function return bool?

Reply:
simply because the rest of the flash code all worked with int returns, which like i noted, I think needs a good cleanup, I agree